### PR TITLE
Parsing and rendering for token trees and forests.

### LINKED ISSUE #5: Token trees
I've been using `Data.Tree` with a little parser to work on the forest structure of the token stream, but now I think that it would be nice to have a custom type like:

```haskell
data TokenTree = TokenNode TagName [Attr] [TokenTree] | ContentTextLeaf Text | ...
type TokenForest = [TokenTree]
```

Or, a lot more abstract (and slightly too flexible in the structure in that it allows to keep tags in leafs):

```haskell
class IsToken t where
    tk :: Prism' t Token

data TokenTree t = TokenNode t [t] | TokenLeaf t
```

Do you have any opinions?  I have some hope that the trivial `IsToken` instance (if you don't want to store anything other than `Token` in the nodes) will have very little impact on runtime, but I haven't checked.

### PR DESCRIPTION
**Pull request by fisx**
Fixes #5.

### CODE REVIEW

**Comment by fisx** [comment]
I'm not 100% convinced by the idea of `IsToken`, because it's a bit clumsy both in writing the code down and (probably) in running it, but for this to be useful to me I need to have something similar.

It would also nice to prevent the "text node with children" run-time error by using a better fitting tree type.  I'm open to following your preference here.

**Comment by fisx** [comment]
The test suite on travis still appears to be missing the tests:

2016-12-28-184553_1600x900_scrot

**Comment by fisx** [comment]
The crash on 7.8.4 seems to be caused by mem exhaustion.  I wonder if I somehow broke travis caching by installing hspec-discover?  Wouldn't see how, but who knows?

File: html-parse.cabal [hunk 1]
```diff
@@ -49,8 +49,9 @@ benchmark bench
 
 test-suite spec
   type:                exitcode-stdio-1.0
-  main-is:             Spec.hs
   hs-source-dirs:      tests
+  main-is:             Spec.hs
+  other-modules:       Text.HTML.ParserSpec, Text.HTML.TreeSpec
```
Comment (bgamari) [inline html-parse.cabal:54]: Ahh, good catch!

File: Text/HTML/Tree.hs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    ) where
+
+import           Data.Monoid
+import           Data.Tree
+import           Text.HTML.Parser
+
+
+-- | If you want to store additional data into your tree, you can define your own token type and
+-- instantiate this class to define a mapping between your type and 'Token'.  See
+-- 'parseTokenForest''.
```
Comment (bgamari) [inline Text/HTML/Tree.hs]: Too many quotes here.

File: Text/HTML/Tree.hs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    | n == n' = Right $ PStack (Node p (reverse ss) : ss') ps
+popParent n pstack
+    = Left $ ParseTokenForestErrorBracketMismatch pstack (Just $ TagClose n)
+
+pushFlatSibling :: t -> PStack t -> PStack t
+pushFlatSibling t (PStack ss ps) = PStack (Node t [] : ss) ps
+
+
+renderTokenForest :: (IsToken t, Show t) => Forest t -> [t]
+renderTokenForest = mconcat . fmap renderTokenTree
```
Comment (bgamari) [inline Text/HTML/Tree.hs]: It doesn't seem to me like this is a "render" operation. It's really more of a "toTokenStream".

Also, I'm not entirely sure I like imposing `Show` just for error handling. Perhaps we can do something else here?

File: Text/HTML/Tree.hs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+renderTokenTree :: (IsToken t, Show t) => Tree t -> [t]
+renderTokenTree = \case
```
Comment (bgamari) [inline Text/HTML/Tree.hs]: I'm not terribly picky, but it does seem like just could just be a plain old equation-style pattern match. Feel free to keep it as-is though.

**Review by bgamari — COMMENTED** [review]
I'm also not entirely sure I like the `IsToken` concept. Couldn't you simply fmap over the `Forest` after parsing?

File: Text/HTML/Tree.hs (hunk 3 above)
Comment (fisx) [reply Text/HTML/Tree.hs]: "toTokenStream" lacks the information from what (forest).  I'll try "tokens{To,From]Forest", let me know if you want me to keep making up better names.

**Comment by fisx** [comment]
Thanks for the review, all very good points.  This is fun!

Sorry about the `IsToken`, it made more sense in the project where it grew up, but you're right, it should not go into this library.

Now that I've changed everything I wonder if I should have used attoparsec for this as well.  I'm tempted to give it a try and see how much more concise the code will get (and how much faster).

But not tonight.

**Comment by fisx** [comment]
I get sporadic errors in the test suite that I'm pretty confident are due to the self-closing tags `<hr>`, `<br>`, ``.  I want to take another look at the spec for that before I fix it.

I'll get back to you.

Btw, the travis error on 7.8 is still there.  Do you have any idea what to make of that?

**Comment by fisx** [comment]
Sorry I'm making so much noise btw :-).  If you changed your mind and would prefer to keep token trees out of this package, I'm still happy to create an `html-parse-tree`.

**Comment by bgamari** [comment]
> Btw, the travis error on 7.8 is still there. Do you have any idea what to make of that?

Sadly no and I can't reproduce the issue locally.

> Sorry I'm making so much noise btw :-). 

Quite alright!

> If you changed your mind and would prefer to keep token trees out of this package, I'm still happy to create an html-parse-tree.

My only reservation was the typeclass; now since it is gone I have no objection to including this.

Thanks for working on this!

**Comment by fisx** [comment]
> Thanks for working on this!

Pleasure!  It's more procrastination than work...  :-)

> > Btw, the travis error on 7.8 is still there. Do you have any idea what to make of that?

> Sadly no and I can't reproduce the issue locally.

I think I've learned a long time ago that travis was lying about the number of available CPUs.  So I'm trying using fewer CPUs (1, to be specific).

I'm not sure mem exhaustion can be the issue here, though.  What causes `SIGABORT` (`-6`)?

**Comment by fisx** [comment]
Now ghc-7.8 doesn't get the deps straight any more.  It gets weirder and weirder.  I'll revert.

I think this issue is unrelated to this PR, though.  Merge anyway?

**Comment by bgamari** [comment]
> I think this issue is unrelated to this PR, though. Merge anyway?

Let's do it!

**Comment by fisx** [comment]
Ok, cool!  Do I need to do anything?

**Comment by bgamari** [comment]
Other than remind me, nope. ;)

**Pull request merged**